### PR TITLE
[Enhancement] Bind un-correlate subquery with his predicate table

### PR DESCRIPTION
**Pull request by Seaven**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Like SQL:
```
select * from t0 join t1 on t0.v2 = t1.v2 where t1.v3 > (select t2.v3 from t2);
```

when finish relation tansform(analyze From-Clause first and analyze Where-Clause later), the plan tree will be:

```
                 Filter
                   |
                 Apply
                /     \
              Join    subquery(t2, must one row)
             /    \
            t0    t1
```

For un-correlate subquery, the result of subquery must one rows, and ApplyNode always transform to Cross-Join. And the Apply-Join node will do join-reorder rule later, it's maybe generate bad plan, because Filter maybe not EQ-Predicate and optimizer can't estimate the selectivity of non-EQ-Predicate, and final will transform to plan:
```
                 Filter
                   |
          Cross Join
                /     \
            Join    subquery(t2, must one row)
             /    \
            t0    t1
```
it's means the result of t0 x t1 maybe is large rows. 

Usually, the predicate of un-correlate subquery has good effect for selectivity, so the subquery join with his outer table which support the un-correlate predicate is better choose, like that:
```
      Filter
         |
      Join
     /       \
   t0   Cross Join
         /         \
        t1        subquery(t2, must one row)
```

so, we support push down apply to do the work

TPCDS-query54: 40s -> 10s

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 79 / 81 (97.53%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownApplyLeftProjectRule.java|22|23|95.65%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownApplyLeftRule.java|26|27|96.30%|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalJoinOperator.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SubqueryTransformer.java|20|20|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalApplyOperator.java|6|6|100.00%|

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 97 / 98 (98.98%)




#### file detail

|   |path|covered line|new line|coverage|not covered line detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownApplyLeftProjectRule.java|17|18|94.44%|[53]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalJoinOperator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SubqueryTransformer.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownApplyLeftRule.java|25|25|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/TaskContext.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/TopDownRewriteTask.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalApplyOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/SubqueryUtils.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 100 / 101 (99.01%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownApplyLeftProjectRule.java|17|18|94.44%|[53]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalJoinOperator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SubqueryTransformer.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownApplyLeftRule.java|28|28|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/TaskContext.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/TopDownRewriteTask.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalApplyOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/SubqueryUtils.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 97 / 99 (97.98%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownApplyLeftProjectRule.java|15|16|93.75%|[53]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownApplyLeftRule.java|27|28|96.43%|[76]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalJoinOperator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SubqueryTransformer.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/TaskContext.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/TopDownRewriteTask.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalApplyOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/SubqueryUtils.java|6|6|100.00%|[]|

**Review by Youngwb — DISMISSED** [review]

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 97 / 99 (97.98%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownApplyLeftProjectRule.java|15|16|93.75%|[53]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownApplyLeftRule.java|27|28|96.43%|[76]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalJoinOperator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SubqueryTransformer.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/TaskContext.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/TopDownRewriteTask.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalApplyOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/SubqueryUtils.java|6|6|100.00%|[]|

**Review by Youngwb — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalApplyOperator.java [hunk 1]
```diff
@@ -50,6 +50,14 @@ public class LogicalApplyOperator extends LogicalOperator {
 
     private final boolean needOutputRightChildColumns;
 
+    /**
+     * Record un-correlation subquery outer table column, use for push down apply node
+     * e.g.
+     * SQL: select * from t1 where t1.v1 > (....);
+     * OuterColumns: t1.v1
+     */
+    private final ColumnRefSet unCorrelationSubqueryOuterColumns;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalApplyOperator.java:59]: Why couldn't use output field?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 2]
```diff
@@ -333,6 +333,16 @@ private void collectAllScanOperators(Memo memo, TaskContext rootTaskContext) {
         rootTaskContext.setAllScanOperators(Collections.unmodifiableList(list));
     }
 
+    private void rewriteSubquery(Memo memo, TaskContext rootTaskContext) {
+        do {
+            rootTaskContext.resetRewriteNum();
+            ruleRewriteIterative(memo, rootTaskContext, RuleSetType.PUSH_DOWN_SUBQUERY);
+        } while (rootTaskContext.hasRewrite());
+        rootTaskContext.resetRewriteNum();
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:341]: needn't

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 3]
```diff
@@ -333,6 +333,16 @@ private void collectAllScanOperators(Memo memo, TaskContext rootTaskContext) {
         rootTaskContext.setAllScanOperators(Collections.unmodifiableList(list));
     }
 
+    private void rewriteSubquery(Memo memo, TaskContext rootTaskContext) {
+        do {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:337]: comment why need to rewrite PUSH_DOWN_SUBQUERY rule exhaustive

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalApplyOperator.java (hunk 1 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalApplyOperator.java:59]: it's different, output means the result of subquery, this field means what columns will do predicate with subquery

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java (hunk 3 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:337]: when multi subquery. like `where v1 < (subquery) and v2 > (subquery)`.

like 
```
Apply
    |
Apply
    |
Project
```

I want to get
```
Project
    |
Apply
    |
Apply
```

But it's hard on our framework

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java (hunk 2 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:341]: Just keep right output state

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalApplyOperator.java (hunk 1 above)
Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/logical/LogicalApplyOperator.java:59]: if so, unCorrelationSubqueryPredicateColumns is more better?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java (hunk 3 above)
Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:337]: what I mean is comment in code.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 97 / 99 (97.98%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownApplyLeftProjectRule.java|15|16|93.75%|[53]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownApplyLeftRule.java|27|28|96.43%|[76]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/RelationTransformer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalJoinOperator.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SubqueryTransformer.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/TaskContext.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/TopDownRewriteTask.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalApplyOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/SubqueryUtils.java|6|6|100.00%|[]|

**Review by kangkaisen — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by Seaven** [comment]
run starrocks_admit_test

**Pull request merged**